java/sage/io/SageFileSource.java

```diff
@@ -0,0 +1,277 @@
+   */
+  public long skip(long n) throws IOException;
+
+  /**
+   * Seek to a specific position relative to the start of the file.
+   *
+   * @param pos The new file position.
+   * @throws IOException If there is an I/O related error or <code>pos</code> is negative.
+   */
+  public void seek(long pos) throws IOException;
+
+  /**
+   * Get the position from which the next byte will be read or written relative to the start of the
+   * file.
+   * <p/>
+   * This method may return -1 if getting the position causes an I/O exception.
+   *
+   * @return The current position.
+   */
+  public long position();
+
+  /**
+   * Get the current length of the file.
+   *
+   * @return The current length of the file.
+   * @throws IOException If there is an I/O related error.
+   */
+  public long length() throws IOException;
+
+  /**
+   * Get the number of bytes remaining in the file.
+   *
+   * @return The number of bytes remaining.
+   * @throws IOException If there is an I/O related error.
+   */
+  public long available() throws IOException;
+
+  /**
+   * Changes the current length of the file.
+   * <p/>
+   * If the current file position is greater than the new length, the file position will be changed
+   * to equal the new file length. If the new file length is greater than the current size of the
+   * file, the file will grow to the new size.
+   *
+   * @param newLength The new file length.
+   * @throws IOException If there is an I/O related error, the file was opened as read-only or this
+   *                     implementation does not support writing.
+   */
+  public void setLength(long newLength) throws IOException;;
+
+  /**
+   * Sync data to disk.
+   *
+   * @throws IOException If there is an I/O related error.
+   */
+  public void sync() throws IOException;
+
+  /**
+   * Flush buffers to disk.
+   * <p/>
+   * This will generally have no effect unless there is a buffer layer present.
+   *
+   * @throws IOException If there is an I/O related error.
+   */
+  public void flush() throws IOException;
+
+  /**
+   * Close file.
+   * <p/>
+   * Ensure that any form of buffering (if any) is flushed before closing the file.
+   *
+   * @throws IOException If there is an I/O related error.
+   */
+  public void close() throws IOException;
+
+  /**
+   * Is this file actively growing?
+   *
+   * @return <code>true</code> if the file is actively growing.
+   */
+  public boolean isActiveFile();
+
+  /**
+   * Is this file opened as read only?
+   *
+   * @return <code>true</code> if the file is read only.
+   */
+  public boolean isReadOnly();
+
+  /**
+   * Execute a command specific to this implementation.
+   *
+   * @param command The command to execute.
+   * @return The result from the command. This could be null. The response is implementation
+   *         specific.
+   * @throws IOException If there is an I/O related error.
+   */
+  public String executeCommand(String command) throws IOException;
```

Comment:
The executeCommand methods seem very odd for a file class. :) It looks like they are just being used to send commands for network files...which makes sense; so in that case, I'd think you would just make them part of the RemoteFileChannel class (and currently they are only used internally within that class, so they wouldn't even need to be public).

I do understand that maybe you put it at this level so then they could push through all the wrappers properly...but as all the non-network cases would just ignore the command, it seems like usage of it would be ambiguous unless you actually knew you were dealing with a RemoteSageFile.  So if that's what you were thinking, maybe then when you need this access later you just create a static method somewhere which can get the RemoteSageFile source from an arbitrary SageFileSource object.

Feel free to explain further, I may be totally missing the point of this. :)